### PR TITLE
Add server option to control total pinned memory size can be used by TRTIS

### PR DESCRIPTION
**Pull request by GuanLuo**

### CODE REVIEW

File: src/core/pinned_memory_manager.cc [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+#include <sstream>
+
+#include "src/core/logging.h"
+#include "src/core/pinned_memory_manager.h"
```
Comment (deadeyegoodwin) [inline src/core/pinned_memory_manager.cc]: This include at top separated by blank line from other includes.

File: src/core/pinned_memory_manager.cc [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    }
+#else
+    *ptr = nullptr;
+    status = Status(
+        RequestStatusCode::INTERNAL,
+        "failed to allocate pinned system memory: " +
+            "TRTIS_ENABLE_GPU is not set"));
+#endif  // TRTIS_ENABLE_GPU
+  }
+  if ((!status.IsOk()) && allow_nonpinned_fallback) {
```
Comment (deadeyegoodwin) [inline src/core/pinned_memory_manager.cc]: Blank line above here

File: src/core/pinned_memory_manager.cc [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+            "unexpected memory address collision, '" + PointerToString(*ptr) +
+                "' has been managed");
+      }
+    }
+    // In either case, we need to adjust back the pinned byte size
+    if ((!status.IsOk()) || (!is_pinned)) {
+      instance_->total_pinned_byte_size_ -= size;
+    }
+  }
+  if ((!status.IsOk()) && (*ptr != nullptr)) {
```
Comment (deadeyegoodwin) [inline src/core/pinned_memory_manager.cc]: Blank line above

File: src/core/pinned_memory_manager.cc [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+        RequestStatusCode::INTERNAL,
+        "failed to allocate pinned system memory: " +
+            "TRTIS_ENABLE_GPU is not set"));
+#endif  // TRTIS_ENABLE_GPU
+  }
+  if ((!status.IsOk()) && allow_nonpinned_fallback) {
+    static bool warning_logged = false;
+    if (!warning_logged) {
+      LOG_WARNING << status.Message()
+                  << ", allocating non-pinned system memory";
```
Comment (deadeyegoodwin) [inline src/core/pinned_memory_manager.cc]: "... pinned memory pool exceeded, falling back to non-pinned system memory".

File: src/core/pinned_memory_manager.h [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+#pragma once
+
+#include <memory>
+
```
Comment (deadeyegoodwin) [inline src/core/pinned_memory_manager.h]: No blank line here, formatter will order them correctly

File: src/core/pinned_memory_manager.h [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  // Options to be checked before allocating pinned memeory
+  struct Options {
+    Options(size_t b = 0) : max_total_byte_size_(b) {}
+
+    size_t max_total_byte_size_;
+  };
+
+  ~PinnedMemoryManager() = default;
+
+  static Status Create(const Options& options);
```
Comment (deadeyegoodwin) [inline src/core/pinned_memory_manager.h]: Some comments on these methods.

File: src/core/trtserver.h [hunk 7]
```diff
@@ -538,6 +538,15 @@ TRTSERVER_EXPORT TRTSERVER_Error* TRTSERVER_ServerOptionsSetStartupModel(
 TRTSERVER_EXPORT TRTSERVER_Error* TRTSERVER_ServerOptionsSetStrictModelConfig(
     TRTSERVER_ServerOptions* options, bool strict);
 
+/// Set the total pinned memory byte size that the server can allocate
+/// in a server options. This option will not affect the allocation conducted
+/// by the backend frameworks.
+/// \param options The server options object.
+/// \param size The total pinned memory byte size.
+/// \return a TRTSERVER_Error indicating success or failure.
+TRTSERVER_EXPORT TRTSERVER_Error* TRTSERVER_ServerOptionsSetTotalPinnedSize(
```
Comment (deadeyegoodwin) [inline src/core/trtserver.h]: TRTSERVER_ServerOptionsSetPinnedMemoryPoolByteSize

File: src/servers/main.cc [hunk 8]
```diff
@@ -282,6 +283,12 @@ std::vector<Option> options_{
      "Name of the model to be loaded on server startup. It may be specified "
      "multiple times to add multiple models. Note that this option will only "
      "take affect if --allow-model-control is true."},
+    {OPTION_TOTAL_PINNED_MEMORY_BYTE_SIZE, "total-pinned-memory-byte-size",
```
Comment (deadeyegoodwin) [inline src/servers/main.cc]: pinned-memory-pool-byte-size

File: src/servers/main.cc [hunk 9]
```diff
@@ -282,6 +283,12 @@ std::vector<Option> options_{
...
+     "The total byte size that can be allocated as pinned system memory. "
+     "If GPU support is enabled, the server will allocate pinned system memory "
+     "to accelerate data transfer between host and devices until it exceeds "
+     "the specified byte size. This option will not affect the allocation "
+     "conducted by the backend frameworks. Default is 0."},
```
Comment (deadeyegoodwin) [inline src/servers/main.cc]: Default is tricky, but I think we should default to a non-zero value. Even on a system with not much CPU memory we should be able to have the default be 256MB.

File: src/servers/main.cc [hunk 10]
```diff
@@ -755,6 +762,7 @@ Parse(TRTSERVER_ServerOptions* server_options, int argc, char** argv)
   VgpuOption tf_vgpu;
   int32_t exit_timeout_secs = 30;
   int32_t repository_poll_secs = repository_poll_secs_;
+  int32_t total_pinned_size = 0;
```
Comment (deadeyegoodwin) [inline src/servers/main.cc]: We need int64_t here so that more than 2GB can be requested.

File: src/core/trtserver.h [hunk 11]
```diff
@@ -538,6 +538,15 @@ TRTSERVER_EXPORT TRTSERVER_Error* TRTSERVER_ServerOptionsSetStartupModel(
...
+    TRTSERVER_ServerOptions* options, unsigned int size);
```
Comment (deadeyegoodwin) [inline src/core/trtserver.h]: Need uint64_t for size to that more than 2GB can be requested.

File: src/core/pinned_memory_manager.cc (hunk 4 above)
Comment (GuanLuo) [reply src/core/pinned_memory_manager.cc]: The reason for not allocating pinned memory will be explained in `status.Message()`, so the new commit only changes here to ", falling back to non-pinned system memory"

**Pull request merged**